precheck sysctl configuration

alexandre-allard opened the pull request:
**Component**: salt

**Context**: 

**Summary**:

**Acceptance criteria**: 


---

Closes: #2853

alexandre-allard committed:
salt: fix install recommended packages

We were installing mandatory instead of
recommended packages in the `recommended.sls`
Salt formula.

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2853-precheck-system-configuration` into
`development/2.6`
* [w/2.7/improvement/2853-precheck-system-configuration](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2853-precheck-system-configuration) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

bert-e commented:
# History mismatch



Merge commit #fbeec939fa8d17f1155bb563ea9ad917eb1528d4 on the integration branch
`w/2.7/improvement/2853-precheck-system-configuration` is merging a branch which is neither the current
branch `improvement/2853-precheck-system-configuration` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/2853-precheck-system-configuration` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on salt/metalk8s/internal/preflight/recommended.sls:
```diff
@@ -4,7 +4,7 @@
   - metalk8s.repo
 
-{%- for pkg_name in kubeadm_preflight.mandatory.packages %}
+{%- for pkg_name in kubeadm_preflight.recommended.packages %}
```
Ohhh, good catch :+1:

TeddyAndrieux commented on salt/metalk8s/internal/preflight/check.sls:
```diff
@@ -0,0 +1,29 @@
+{%- from "metalk8s/map.jinja" import kubeadm_preflight with context %}
+
+{%- set status = {'error': False} %}
+{%- for sysctl, expected_value in kubeadm_preflight.mandatory.sysctl_values.items() %}
+  {%- set value = salt.sysctl.get(sysctl).strip('"') %}
+Check sysctl {{ sysctl }} value is {{ expected_value }}:
+  test.configurable_test_state:
+    - changes: False
+    - comment: {{ sysctl }} value is {{ value }}
+  {%- if value|int != expected_value|int %}
+  {%- do status.update({'error': True}) %}
+    - result: False
+  {%- endif %}
+{%- endfor %}
+
+All sysctl parameters are well configured:
+  test.configurable_test_state:
+    - changes: False
+    - comment: |-
+{%- if status.error %}
+        Some sysctl parameters have been set during installation, but
+        their values are not taken in account, please check the sysctl
+        configuration to ensure MetalK8s one have precedence.
+        See above to find parameters with wrong values.
+    - result: False
+    - failhard: True
+{%- else %}
+        Success!
+{%- endif %}
```
Does it works ? For me, jinja is rendered at the same time as the one in `mandatory.sls` so you will not get up to date values
I think you need to use slot here (and btw slot will not be enough you need a custom module/state)
I would say a simple module to compare sysctl that just return True or false and then a slot on `result` in `test.configurable_test_state` (like we do in salt minion running state)

Also to me this state could be part of  `mandatory.sls` no ?

alexandre-allard replied:
Yes, you're right, I just saw this when testing, I'm currently changing it.
I'm not sure it should be in `mandatory.sls`, it depends if we want it to be callable without doing any change which was the idea (so we can launch this on a regular basis through beacon/cron/whatever and send alert when it fails).

TeddyAndrieux replied:
Right, yes it can be useful but to me, it's not the goal of a salt state if we want something like that it should be a module directly instead I think.
Maybe you should create a simple module `metalk8s_checks.py` with a function for sysctl that basically do exactly the same as this state and then in mandatory state you just do a call to this module, what do you think ? (and then in the futur we could imagine adding other checks in this modules if needed)

sayf-eddine-scality replied:
slowly but steadily building a platformcheck clone ;p. btw did you consider having the mandatory sysctl settings in /etc/sysctl.conf directly ? that way they re not overwritten (I guess)

alexandre-allard replied:
@TeddyAndrieux haha, that's what I'm currently doing. :)
@sayf-eddine-scality we could do that also, it works for sure.

alexandre-allard replied:
This is the order the sysctl are read (it's like every following directories are merged, first definition of a specific filename overwrite the others, then configuration files are applied by ASCII order, with the last one, always being `/etc/sysctl.conf`):
```
/run/sysctl.d/*.conf
/etc/sysctl.d/*.conf
/usr/local/lib/sysctl.d/*.conf
/usr/lib/sysctl.d/*.conf
/lib/sysctl.d/*.conf
/etc/sysctl.conf
```
The last value defined overwrite all the previous one.

TeddyAndrieux replied:
For me, we do not want to use `/etc/sysctl.conf` so that if one day we want to clean up a server it's really simpler to just remove our files  (cc @sayf-eddine-scality )

sayf-eddine-scality replied:
I am just suggesting u chose.
That is not the first time we face such an issue, we simply don't have a lot of control over drop_in files. either
- we setup our own file which is recommended but we risk ppl or other tools to take precedence
- or we simply add it directly in the root file and maybe add `# added by metalk8s` to be able to clean it later.

alexandre-allard committed:
salt: add postcheck on mandatory sysctl values

We now ensure that the sysctl values we set are
taken in account, if not, we fail early in the
process.

Refs: #2853

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/2853-precheck-system-configuration` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `improvement/2853-precheck-system-configuration`
* Rebase `improvement/2853-precheck-system-configuration` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# History mismatch



Merge commit #83ef725a6f8aa2f8bf192cf694fc3880c243ea56 on the integration branch
`w/2.7/improvement/2853-precheck-system-configuration` is merging a branch which is neither the current
branch `improvement/2853-precheck-system-configuration` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/2853-precheck-system-configuration` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2853-precheck-system-configuration` into
`development/2.6`
* [w/2.7/improvement/2853-precheck-system-configuration](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2853-precheck-system-configuration) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @alexandre-allard-scality

* @jbertran

* @sayf-eddine-scality

* @teddyandrieux

* @nicolast

* @lucieleonard

* @nootal

* @monpote

* @cmonfort

* @slaperche-scality

* @tcarmet

* @carlito-scality

* @thomasdanan

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

alexandre-allard committed:
salt: add unit tests for metalk8s_checks module

Refs: #2853

gdemonet commented on salt/_states/metalk8s_checks.py:
```diff
@@ -0,0 +1,35 @@
+    return __virtualname__
+
+
+def module_run(name, **kwargs):
```
Why wrap `module.run`? Since we have a custom state module and the corresponding execution module, why not just call it directly?

Something like:

```sls
Check sysctl overrides are properly applied:
  metalk8s_checks.sysctl:
    - params: {{ kubeadm_preflight.mandatory.sysctl_values | json }}
    - require:
      - cmd: Apply sysctl configuration files
```

And:

```python
def sysctl(name, params, **kwargs):
  ret = {
    "name": name,
    "changes": {},
    "result": True,
    "comment": "All sysctl settings are correctly configured."
  }
  error_msg = __salt__["metalk8s_checks.sysctl"](params, raises=False)
  if error_msg:
    ret["result"] = False
    ret["comment"] = error_msg
  return ret
```

alexandre-allard replied:
It's what I did first, but it was to have something more generic.
To have a state able to run some "test" execution modules if we need to add some more tests later.
Also, wrapping module.run allow to specify multiple exec modules in one state, such as:
```
Check stuff:
  metalk8s_checks.module_run:
    - metalk8s_checks.sysctl:
        - params: {{ kubeadm_preflight.mandatory.sysctl_values | json }}
    - metalk8s_checks.some_other_tests:
        - some_args: 123
```
But maybe we don't need this and we prefer to define checks one by one.

gdemonet replied:
Oh. I see the idea, just wondering what benefit we have (outside of less YAML to write):
- if one module from the list fails (raises a `CheckError`), we'll only get output for this one
- the state ID for the `module.run` will not be specific, so we'll need to rely on the comment to understand what was being checked and what failed

If we really want something able to handle multiple checks (not sure it's worth the effort, TBH), we'll need to do a bit more than just wrapping `module.run` and catching the first `CheckError` that comes our way.
If we really want something able to handle any check, then maybe we don't need to wrap `module.run`, simply:

```python
def run(name, mod, **kwargs):
    ret = {
        "name": name,
        "changes": {},
        "result": True,
        "comment": "Check succeeded.",
    }

    try:
        comment = __salt__[mod](**kwargs)
    except KeyError:
        ret["result"] = False
        ret["comment"] = "Module `{}` does not exist.".format(mod)
    except CheckError as exc:
        ret["result"] = False
        ret["comment"] = str(exc)

    if comment:
        ret["comment"] = comment

    return ret
```

gdemonet replied:
And if we really want to ensure this only runs known checks, we can restrict it with:

```python
def run(name, func, **kwargs):
    ret = { ... }

    try:
        comment = __salt__['metalk8s_checks.{}'.format(func)](**kwargs)
    except ...
```

TeddyAndrieux replied:
Even if we only do one call per module.run at least here we have only one custom state to do all checks instead of writing a states + a module for each check we want to do. 
And btw a state for check does not really make sense at the end, a state should do some modification on the system and not only check stuff, we could imagine in the future having a "tool" that basically just run the module directly.

But no opinion about call state `module.run` vs call our module directly. Btw if we stay will `module.run` at least we can consider this state work the same as all other "classic module.run" states but it's not really an argument because anyway our states will be really simple so .... as you want

alexandre-allard replied:
done

gdemonet reviewed (commented):
most of it looks good, just a bit confused about why we want to wrap "module.run"

TeddyAndrieux commented on salt/_states/metalk8s_checks.py:
```diff
@@ -0,0 +1,35 @@
+
+
+def module_run(name, **kwargs):
+    """Wrap `module.run` state, pretty print the error and fail gracefully if
+    the module raises a `CheckError` exception.
```
Minor: Mention also that this state should really do NOT changes anything on the host it's pure "checks"

alexandre-allard replied:
Done, I've also put a little explanation why we're using a state to do this.

TeddyAndrieux commented on salt/_modules/metalk8s_checks.py:
```diff
@@ -0,0 +1,41 @@
+        current_value = __salt__["sysctl.get"](key)
+        if current_value != str(value):
+            errors.append(
+                "Incorrect value for '{0}', expecting '{1}' but "
```
Maybe just add the fact we talk about sysctl here even if it's quite clear with the value

```suggestion
                "Incorrect value for sysctl '{0}', expecting '{1}' but "
```

alexandre-allard replied:
done

TeddyAndrieux commented on salt/metalk8s/internal/preflight/mandatory.sls:
```diff
@@ -38,8 +38,22 @@ Set sysctl {{ item }} value to {{ value }}:
+      - cmd: Apply sysctl configuration files
 {%- endfor %}
 
+{#- This is needed to ensure values loaded from files match in-memory ones. #}
```
Minor: We do not really care, but here it's a comment about YAML not jinja so to me it should be a YAML comment :smiley:

alexandre-allard replied:
done

TeddyAndrieux commented on salt/tests/unit/modules/test_metalk8s_checks.py:
```diff
@@ -0,0 +1,60 @@
+        """
+        Tests the return of `sysctl` function
+        """
+        module_run_mock = MagicMock(side_effect=data.get)
```
Minor:

```suggestion
        sysctl_get_mock = MagicMock(side_effect=data.get)
```

alexandre-allard replied:
done

TeddyAndrieux reviewed (commented):
LGTM just a few minor comments

bert-e commented:
# History mismatch



Merge commit #8e5872037195ff61fbc30c7d67a5550b03f7924d on the integration branch
`w/2.7/improvement/2853-precheck-system-configuration` is merging a branch which is neither the current
branch `improvement/2853-precheck-system-configuration` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/2853-precheck-system-configuration` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2853-precheck-system-configuration` into
`development/2.6`
* [w/2.7/improvement/2853-precheck-system-configuration](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2853-precheck-system-configuration) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

alexandre-allard commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/5faa1b1936bf7e71c9516b5617410edc041fcbed) did not succeed in branch w/2.7/improvement/2853-precheck-system-configuration.




*The following options are set:* **approve**

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/2853-precheck-system-configuration` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `improvement/2853-precheck-system-configuration`
* Rebase `improvement/2853-precheck-system-configuration` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #dcf425c5c2ee73e7b352578609a0fc436c81339c on the integration branch
`w/2.7/improvement/2853-precheck-system-configuration` is merging a branch which is neither the current
branch `improvement/2853-precheck-system-configuration` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/2853-precheck-system-configuration` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2853-precheck-system-configuration` into
`development/2.6`
* [w/2.7/improvement/2853-precheck-system-configuration](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2853-precheck-system-configuration) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
